src/AggregateFunctions/SingleValueData.cpp

```diff
@@ -0,0 +1,1419 @@
+        return true;
+    }
+    else
+        return false;
+}
+
+void SingleValueDataGeneric::insertResultInto(IColumn & to) const
+{
+    if (has())
+        to.insert(value);
+    else
+        to.insertDefault();
+}
+
+void SingleValueDataGeneric::write(WriteBuffer & buf, const ISerialization & serialization) const
+{
+    if (!value.isNull())
+    {
+        writeBinary(true, buf);
+        serialization.serializeBinary(value, buf, {});
+    }
+    else
+        writeBinary(false, buf);
+}
+
+void SingleValueDataGeneric::read(ReadBuffer & buf, const ISerialization & serialization, Arena *)
+{
+    bool is_not_null;
+    readBinary(is_not_null, buf);
+
+    if (is_not_null)
+        serialization.deserializeBinary(value, buf, {});
+}
+
+bool SingleValueDataGeneric::isEqualTo(const IColumn & column, size_t row_num) const
+{
+    return has() && value == column[row_num];
+}
+
+bool SingleValueDataGeneric::isEqualTo(const DB::SingleValueDataBase & other) const
+{
+    auto const & to = assert_cast<const Self &>(other);
+    return has() && to.value == value;
+}
+
+void SingleValueDataGeneric::set(const IColumn & column, size_t row_num, Arena *)
+{
+    column.get(row_num, value);
+}
+
+void SingleValueDataGeneric::set(const SingleValueDataBase & other, Arena *)
+{
+    auto const & to = assert_cast<const Self &>(other);
+    if (other.has())
+        value = to.value;
+}
+
+bool SingleValueDataGeneric::setIfSmaller(const IColumn & column, size_t row_num, Arena * arena)
+{
+    if (!has())
+    {
+        set(column, row_num, arena);
+        return true;
+    }
+    else
+    {
+        Field new_value;
+        column.get(row_num, new_value);
+        if (new_value < value)
+        {
+            value = new_value;
+            return true;
+        }
+        else
+            return false;
+    }
+}
+
+bool SingleValueDataGeneric::setIfSmaller(const SingleValueDataBase & other, Arena *)
+{
+    auto const & to = assert_cast<const Self &>(other);
+    if (to.has() && (!has() || to.value < value))
+    {
+        value = to.value;
+        return true;
+    }
+    else
+        return false;
+}
+
+bool SingleValueDataGeneric::setIfGreater(const IColumn & column, size_t row_num, Arena * arena)
+{
+    if (!has())
+    {
+        set(column, row_num, arena);
+        return true;
+    }
+    else
+    {
+        Field new_value;
+        column.get(row_num, new_value);
+        if (new_value > value)
+        {
+            value = new_value;
+            return true;
+        }
+        else
+            return false;
+    }
+}
+
+bool SingleValueDataGeneric::setIfGreater(const SingleValueDataBase & other, Arena *)
+{
+    auto const & to = assert_cast<const Self &>(other);
+    if (to.has() && (!has() || to.value > value))
+    {
+        value = to.value;
+        return true;
+    }
+    else
+        return false;
+}
+
+void generateSingleValueFromTypeIndex(TypeIndex idx, SingleValueDataBaseMemoryBlock & data)
```

Comment:
this does look like a mix of variant and inheritance 
probably it could be done with just variant, but the code looks okay IMO

Reply:
Probably possible, but there is a couple of things making it harder:
* std::variant with template classes (SingleValueDataNumeric<T>).
* It adds an extra storage to hold the alternative type. It might be possible to recover this by removing the base class and implementing all methods everywhere. Not sure which option is worse